tools/analyze_results.py

```diff
@@ -0,0 +1,208 @@
+        annotation (dict): Ground truth annotations where keys of
+             annotations are:
+
+            - `bboxes`: numpy array of shape (n, 4)
+            - `labels`: numpy array of shape (n, )
+            - `bboxes_ignore` (optional): numpy array of shape (k, 4)
+            - `labels_ignore` (optional): numpy array of shape (k, )
+
+        Returns:
+            float: mAP
+    """
+
+    # use only bbox det result
+    if isinstance(det_result, tuple):
+        bbox_det_result = [det_result[0]]
+    else:
+        bbox_det_result = [det_result]
+    # mAP
+    iou_thrs = np.linspace(
+        .5, 0.95, int(np.round((0.95 - .5) / .05)) + 1, endpoint=True)
+    mean_aps = []
+    for thr in iou_thrs:
+        mean_ap, _ = eval_map(bbox_det_result, [annotation], iou_thr=thr)
+        mean_aps.append(mean_ap)
+    return sum(mean_aps) / len(mean_aps)
+
+
+class ResultVisualizer(object):
+    """Display and save evaluation results.
+
+    Args:
+        dataset (Dataset): A PyTorch dataset.
+        results (pickle object): pickle object from test results pkl file
+        topk (int): Number of the highest topk and
+            lowest topk after evaluation index sorting. Default: 20
+        show (bool): Whether to show the image. Default: True
+        wait_time (float): Value of waitKey param. Default: 0.
+        show_dir (str or None): The filename to write the image.
+            Default: 'work_dir'
+    """
+
+    def __init__(self, show=False, wait_time=0):
+        self.show = show
+        self.wait_time = wait_time
+
+    def _save_image_gts_results(self, dataset, results, mAPs, out_dir=None):
+        mmcv.mkdir_or_exist(out_dir)
+
+        for mAP_info in mAPs:
+            index, mAP = mAP_info
+            data_info = dataset.prepare_train_img(index)
+
+            # calc save file path
+            filename = data_info['filename']
+            if data_info['img_prefix'] is not None:
+                filename = osp.join(data_info['img_prefix'], filename)
+            else:
+                filename = data_info['filename']
+            fname, name = osp.splitext(osp.basename(filename))
+            save_filename = fname + '_' + str(round(mAP, 3)) + name
+            out_file = osp.join(out_dir, save_filename)
+            imshow_gt_det_bboxes(
+                data_info['img'],
+                data_info,
+                results[index],
+                dataset.CLASSES,
+                show=self.show,
+                wait_time=self.wait_time,
+                out_file=out_file)
+
+    def evaluate_and_show(self,
+                          dataset,
+                          results,
+                          topk=20,
+                          show_dir='work_dir',
+                          eval_fn=None):
+        """Evaluate and show results.
+
+        Args:
+            dataset (Dataset): A PyTorch dataset.
+            results (pickle object): pickle object from test results pkl file
```

Comment:
pickle object is not a valid type. It is the original type, for example, a list of tuple or some thing.